### PR TITLE
BC-2161 - copy modal makeover

### PR DESCRIPTION
**Pull request by OliverHappe**
# Short Description

Refactor CopyProcess and CopyResult components to be easier to understand and better to maintain.

## Links to Ticket and related Pull-Requests

- https://ticketsystem.dbildungscloud.de/browse/BC-2161
https://github.com/hpi-schul-cloud/schulcloud-server/pull/3592
## Checklist before merging

- [x] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [x] PO: Any deviation from requirements was agreed with Product-Owner / ticket author / support-team

> Notice: Please keep this Pull-Request as a Draft (or add WIP label), until it is ready to be reviewed

### CODE REVIEW

**Comment by OliverHappe** [comment]
next step: 
 * move business logic regarding copying either to the parent component (room-dashboard to room-details) or create a second store for different use cases

File: src/store/copy.ts [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+		}
+		if (payload.elements === undefined) {
+			this.copyResultFailedItems = [];
+			return;
+		}
+
+		const isFeedbackParent: (type: CopyApiResponseTypeEnum) => boolean = (
+			type
+		) => {
+			if (type === CopyApiResponseTypeEnum.Course) return true;
```
Comment (MartinSchuhmacher) [inline src/store/copy.ts:110]: how about linking with || and just return the boolean for the comparison?

File: src/store/copy.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+			if (type === CopyApiResponseTypeEnum.Lesson) return true;
+			if (type === CopyApiResponseTypeEnum.Task) return true;
+			if (type === CopyApiResponseTypeEnum.LernstoreMaterialGroup) return true;
+			return false;
+		};
+
+		const isFeedbackChild: (type: CopyApiResponseTypeEnum) => boolean = (
+			type: CopyApiResponseTypeEnum
+		) => {
+			if (type === CopyApiResponseTypeEnum.LessonContentEtherpad) return true;
```
Comment (MartinSchuhmacher) [inline src/store/copy.ts:120]: see above

File: src/serverApi/v3/api.ts [hunk 3]
```diff
@@ -244,8 +244,13 @@ export enum CopyApiResponseTypeEnum {
     FileGroup = 'FILE_GROUP',
```
Comment (virgilchiriac) [inline src/serverApi/v3/api.ts:244]: TODO: check this file is no different than when running lastest main on server

File: src/components/copy-result-modal/CopyResultModal.vue [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+				>
+					<div class="alert_text mr-2">
+						{{ $t("components.molecules.copyResult.failedCopy") }}
+					</div>
+				</v-alert>
+				<div v-if="hasNotification" data-testid="copy-result-notifications">
+					<v-alert type="warning" :icon="mdiInformation" text border="left">
+						<div class="alert_text mr-2">
+							<div v-if="hasFailedGeogebraElement" data-testid="geogebra">
+								<strong>Geogebra</strong> &middot;
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/copy-result-modal/CopyResultModal.vue:43]: Thanks for the idea. That's an alternative. But not necessarily better.

File: src/components/copy-result-modal/types/CopyResultItem.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ *  Enum.LEARNSTORE_MATERIAL_GROUP > Enum.LEARNSTORE_MATERIAL (leaf-type we need)
+ *  Enum.TASK_GROUP > Enum.TASK
+ *
+ * Kurse (Enum.COURSE):
+ *  Enum.COURSE_GROUP -> not-implemented
+ *
+ * Board
+ *  Enum.BOARD -> Merged Kinder von TASK und LESSON
+ *
+ */
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/copy-result-modal/types/CopyResultItem.ts:27]: Removed unneeded comments (commit will follow).

File: src/components/copy-result-modal/types/CopyResultItemElement.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import { CopyApiResponseTypeEnum } from "../../../serverApi/v3";
+
+export interface CopyResultItemElement {
+	type: CopyApiResponseTypeEnum; // made up values just to transport my thoughts
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/copy-result-modal/types/CopyResultItemElement.ts:4]: Removed unneeded comment (commit will follow).

File: src/components/copy-result-modal/CopyResultModal.vue [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+				return this.$t("components.molecules.copyResult.title.loading");
+			} else if (this.copyResultStatus !== undefined) {
+				switch (this.copyResultStatus) {
+					case "success":
+						return this.$t("components.molecules.copyResult.title.success");
+
+					case "partial":
+						return this.$t("components.molecules.copyResult.title.partial");
+
+					case "failure":
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/copy-result-modal/CopyResultModal.vue:158]: To clearify, what that it was not forgotten but decided how to handle case "failure".

File: src/serverApi/v3/api.ts (hunk 3 above)
Comment (MartinSchuhmacher) [reply src/serverApi/v3/api.ts:244]: but there should be a difference as we have the according server PR on the same branch

File: src/components/templates/TasksDashboardMain.vue [hunk 8]
```diff
@@ -6,7 +6,7 @@
 	>
 		<div slot="header">
 			<div>
-				<h1 class="text-h3">{{ $t("common.words.tasks") }}</h1>
+				<h1 class="text-h3">{{ $t("common.words.tasks") }} Main</h1>
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/templates/TasksDashboardMain.vue:9]: Totally Main! ;-)

File: src/store/copy.ts (hunk 1 above)
Comment (hoeppner-dataport) [reply src/store/copy.ts:110]: The textlayout was chosen intentionally (as far as I now).

File: src/store/copy.ts (hunk 2 above)
Comment (hoeppner-dataport) [reply src/store/copy.ts:120]: too ;-)

File: src/components/copy-result-modal/CopyResultModal.unit.ts [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+		it("should show coursegroup notification if failed coursegroup element exists", () => {
+			const courseGroupNotification = setupWrapper().find(
+				`[data-testid="copy-result-notifications"] [data-testid="coursegroups"]`
+			);
+
+			expect(courseGroupNotification.isVisible()).toBe(true);
+		});
+
+		it("should show coursegroup notification if file error element exists", () => {
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/copy-result-modal/CopyResultModal.unit.ts]: wrong description text. fixed.

File: src/components/copy-result-modal/CopyResultModal.unit.ts [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+				copyResultStatus: undefined,
+			});
+			const headline = wrapper.find("h2").element.textContent;
+
+			expect(headline).toContain(
+				wrapper.vm.$i18n.t("components.molecules.copyResult.title.loading")
+			);
+		});
+
+		it("should show success-title when ", () => {
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/copy-result-modal/CopyResultModal.unit.ts]: descriptions completed. fixed.

File: src/components/copy-result-modal/CopyResultModal.unit.ts [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+			dialog.vm.$emit("dialog-closed");
+
+			const emittedEvents = wrapper.emitted();
+			const dialogClosedEvents = emittedEvents["dialog-closed"];
+			expect(dialogClosedEvents?.length).toBe(1);
+		});
+	});
+
+	describe("status alerts", () => {
+		it("should render success alert  if status is success", async () => {
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/copy-result-modal/CopyResultModal.unit.ts]: fixed.

**Review by MartinSchuhmacher — APPROVED** [review]

**Pull request merged**